### PR TITLE
Ignore tag's case

### PR DESCRIPTION
**Pull request by Kdecherf**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | somewhat?
| Deprecations? | no
| Tests pass?   | -
| Documentation | no
| Translation   | no
| Fixed tickets | #2502 
| License       | MIT

Store tag labels in lowercase in order to prevent case-sensitive duplicates.

Non-lowercase tags already existing before this PR should still work as usual but any new tag will be saved in lowercase.

### CODE REVIEW

**Comment by j0k3r** [comment]
Is test on MySQL are green locally?

**Comment by Kdecherf** [comment]
@j0k3r will check next week

**Comment by Kdecherf** [comment]
Well... After checking the failing test I think I'm hitting a case sensitivity issue with databases. I will rework the PR.

**Comment by Kdecherf** [comment]
I updated the test for WallabagV1Import. Please review carefully @j0k3r @nicosomb @tcitworld 

Reminder: this PR removes the ability to have upper chars in tag labels.

**Comment by nicosomb** [comment]
Do we need a migration to update current tags?

File: src/Wallabag/CoreBundle/Helper/ContentProxy.php [hunk 1]
```diff
@@ -146,4 +146,52 @@ private function validateContent(array $content)
...
+
+    /**
+     * Assign some tags to an entry.
+     *
+     * @param Entry        $entry
+     * @param array|string $tags          An array of tag or a string coma separated of tag
+     * @param array        $entitiesReady Entities from the EntityManager which are persisted but not yet flushed
+     *                                    It is mostly to fix duplicate tag on import @see http://stackoverflow.com/a/7879164/569101
+     */
+    public function assignTagsToEntry(Entry $entry, $tags, array $entitiesReady = [])
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Helper/ContentProxy.php]: Hum, in 2.3 `assignTagsToEntry` has moved to `Wallabag\CoreBundle\Helper\TagsAssigner`.
Please update the correct one 🙂

Comment (Kdecherf) [inline src/Wallabag/CoreBundle/Helper/ContentProxy.php]: Well I blindly rebased my work without seeing that the whole method was missing...

**Comment by Kdecherf** [comment]
I need to add a specific test case for checking backward compatibility.

**Comment by j0k3r** [comment]
Go ahead 🙂

**Comment by Kdecherf** [comment]
I broke something really hard while trying to fix some tests. It's way more
complicated than I initially thought when dealing with same tags with
different cases. The better way will be to migrate all previous tags in
lowercase. I'll handle that on the next week.

Sorry for the delay :(

Sent from my phone

**Comment by Kdecherf** [comment]
I am going to resume work on this PR this evening.

**Comment by Kdecherf** [comment]
The migration of tags is mostly done but I found another issue regarding unicode characters: PHP function `strtolower()` and SQL function `LOWER()` don't handle them. In this case wallabag can't merge `héllo` and `hÉllo` tags together.

Should we improve the support of unicode chars on the app side or should we just convert label like slug instead?

**Comment by Kdecherf** [comment]
Well... Correction: only SQLite has issues with LOWER-ing unicode characters.

**Comment by Kdecherf** [comment]
Please check the message of the commit introducing migration for detail. Note also that SQLite is unsupported by this migration.

Extensive test is required for migrating existing databases on MySQL and PostgreSQL.

File: app/DoctrineMigrations/Version20170719231144.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            if (count($ids) > 0) {
+                // Merge tags
+                $this->addSql("
+                    UPDATE '".$this->getTable('entry_tag')."'
+                    SET    tag_id = ".$newId."
+                    WHERE  tag_id IN (".implode(',', $ids).")"
+                );
+
+                // Delete unused tags
+                $this->addSql("
```
Comment (j0k3r) [inline app/DoctrineMigrations/Version20170719231144.php]: What if the tag is associated to some entries?
We should handle the relation table `entry_tag` too 😕

Comment (Kdecherf) [inline app/DoctrineMigrations/Version20170719231144.php]: This is the goal of the SQL block just above it

Comment (j0k3r) [inline app/DoctrineMigrations/Version20170719231144.php]: :rofl: true.

**Comment by j0k3r** [comment]
Could you run a php-cs-fixer? And I think we'll be good to go :slightly_smiling_face:

**Comment by j0k3r** [comment]
👏

**Pull request merged**